server/filestore.go

```diff
@@ -4517,55 +4524,93 @@ func (fs *fileStore) enforceMsgPerSubjectLimit(fireCallback bool) {
 		// Rebuild fs state too.
 		fs.rebuildStateLocked(nil)
 		// Need to redo blocks that need attention.
-		needAttention = make(map[string]*psi)
+		needAttention.Empty()
+		fblk, lblk = uint32(math.MaxUint32), uint32(0)
 		fs.psim.IterFast(func(subj []byte, psi *psi) bool {
 			if psi.total > maxMsgsPer {
-				needAttention[string(subj)] = psi
+				needAttention.Insert(subj, psi.total)
+				if psi.fblk < fblk {
+					fblk = psi.fblk
+				}
+				if psi.lblk > lblk {
+					lblk = psi.lblk
+				}
 			}
 			return true
 		})
 	}
 
+	// If nothing to do then stop.
+	if fblk == math.MaxUint32 {
+		return
+	}
+
 	// Collect all the msgBlks we alter.
 	blks := make(map[*msgBlock]struct{})
 
 	// For re-use below.
 	var sm StoreMsg
 
-	// Walk all subjects that need attention here.
-	for subj, info := range needAttention {
-		total, start, stop := info.total, info.fblk, info.lblk
-
-		for i := start; i <= stop; i++ {
-			mb := fs.bim[i]
-			if mb == nil {
-				continue
+	done := make(chan struct{})
+	defer close(done)
+	i := fblk
+	go func() {
+		t := time.NewTicker(10 * time.Second)
+		started := time.Now()
+		var printed bool
+		for {
+			select {
+			case <-t.C:
```

Comment:
Could we check t one done case and avoid printed bool?

Reply:
I would remove it myself..

Or if we feel strongly we want to keep it in, try to figure out a cleaner way to not have the printed bool.

Reply:
OK, have removed it for now. We can always add it back later if we still find pathological cases where it's slow.